### PR TITLE
sql: treat CREATE TYPE's options as data types

### PR DESCRIPTION
**Pull request by sploiselle**
Had an idea to make type name aliases available to `CREATE TYPE`'s option––is this worth pursuing, or does clamping down the `with_options` to being data types potentially too restrictive in the future?

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5132)

### CODE REVIEW

**Review by benesch — APPROVED** [review]
Seems smart to me!

**Comment by benesch** [comment]
We should probably get rid of `smallint`, `varchar`, and `char`, now that I think about it. It is wrong from a pgcompat perspective to have them aliased the way that they are right now.

**Comment by benesch** [comment]
(That's a general musing—not something to address in this PR!)

**Comment by sploiselle** [comment]
@benesch @JLDLaughlin Ready for actual review

File: src/sql/src/normalize.rs [hunk 1]
```diff
@@ -296,23 +300,18 @@ pub fn create_statement(scx: &StatementContext, mut stmt: Statement) -> Result<S
...
-                    }
-                    SqlOption::Value {
-                        name,
-                        value: Value::String(val),
-                    } => {
-                        *option = SqlOption::ObjectName {
-                            name: name.clone(),
-                            object_name: resolve_item(&ObjectName::unqualified(&val))?,
-                        }
+                    SqlOption::DataType { data_type, .. } => {
```
Comment (JLDLaughlin) [inline src/sql/src/normalize.rs:306]: 🎉

File: src/sql/src/plan/statement.rs [hunk 2]
```diff
@@ -971,12 +971,28 @@ fn handle_create_type(
...
-            Some(SqlOption::Value {
-                value: Value::String(val),
-                ..
-            }) => ObjectName::unqualified(&val),
-            Some(SqlOption::ObjectName { object_name, .. }) => object_name,
-            Some(_) => bail!("{} must be a string or identifier", key),
+            Some(SqlOption::DataType { data_type, .. }) => match data_type {
+                DataType::Other { name, typ_mod } => {
+                    if !typ_mod.is_empty() {
+                        bail!(
```
Comment (JLDLaughlin) [inline src/sql/src/plan/statement.rs:977]: Why can't we accept type modifiers here?

**Review by JLDLaughlin — APPROVED** [review]
this looks great to me!

File: src/sql/src/plan/statement.rs (hunk 2 above)
Comment (sploiselle) [reply src/sql/src/plan/statement.rs:977]: At the highest level, we only store a type's OID, which isn't correlated to any type modifiers.

The only type modifier we actually use is the one on `numeric`/`decimal`, to specify the precision and scale; however, all `numeric` types share the same OID, which means no matter what you specify you end up with the default value––I have this tracked in #4884.

**Review by benesch — APPROVED** [review]
Still LGTM!

**Pull request merged**